swarm/storage/pushsync/pusher_test.go

```diff
@@ -0,0 +1,414 @@
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package pushsync
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/swarm/chunk"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+	colorable "github.com/mattn/go-colorable"
+)
+
+var (
+	loglevel = flag.Int("loglevel", 3, "verbosity of logs")
+)
+
+func init() {
+	flag.Parse()
+	log.PrintOrigins(true)
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(colorable.NewColorableStderr(), log.TerminalFormat(true))))
+}
+
+// loopback implements PubSub as a central subscription engine,
+// ie a msg sent is received by all handlers registered for the topic
+type loopBack struct {
+	async    bool
+	addr     []byte
+	handlers map[string][]func(msg []byte, p *p2p.Peer) error
+}
+
+func newLoopBack(async bool) *loopBack {
+	return &loopBack{
+		async:    async,
+		addr:     make([]byte, 32),
+		handlers: make(map[string][]func(msg []byte, p *p2p.Peer) error),
+	}
+}
+
+// Register subscribes to a topic with a handler
+func (lb *loopBack) Register(topic string, _ bool, handler func(msg []byte, p *p2p.Peer) error) func() {
+	lb.handlers[topic] = append(lb.handlers[topic], handler)
+	return func() {}
+}
+
+// Send publishes a msg with a topic and directly calls registered handlers with
+// that topic
+func (lb *loopBack) Send(to []byte, topic string, msg []byte) error {
+	if lb.async {
+		go lb.send(to, topic, msg)
+		return nil
+	}
+	return lb.send(to, topic, msg)
+}
+
+func (lb *loopBack) send(to []byte, topic string, msg []byte) error {
+	p := p2p.NewPeer(enode.ID{}, "", nil)
+	for _, handler := range lb.handlers[topic] {
+		if err := handler(msg, p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// BaseAddr needed to implement PubSub interface
+func (lb *loopBack) BaseAddr() []byte {
+	return lb.addr
+}
+
+// testPushSyncIndex mocks localstore and provides subscription and setting synced status
+// it implements PushSyncIndex interface
+type testPushSyncIndex struct {
```

Comment:
the tests in this package show no need to mock localstore IMO. there's no tests here that require a lean mock due to some testing with scale. I'd rather use the localstore directly, with the `SubscribePush` and all of the aforementioned methods that semantically fit into the chunk store interface to be put into that interface

Reply:
i disagree.  testing with lean interfaces is cleaner and easier to understand,  plus we will soon need push sync on delete functionality which will need this

Reply:
ok. have it your way. i still disagree. let's see what @janos has to say about this once he gets back